sdk/include/platform/sunburst/platform-ethernet.hh

```diff
@@ -0,0 +1,662 @@
+				}
+				drop_error_frame();
+				continue;
+			}
+
+			// Reset QMU RXQ frame pointer to zero and start DMA transfer
+			// operation.
+			register_write(RegisterOffset::RXFDPR, 0x4000);
+			register_set(RegisterOffset::RXQCR, StartDmaAccess);
+
+			// Start receiving via SPI.
+			uint8_t cmd = 0b10 << 6;
+			set_gpio_output_bit(GpioPin::EthCsPin, false);
+			spi()->blocking_write(&cmd, 1);
+
+			// Initial words are RXFHSR and RXFHBCR which we have already know
+			// the value.
+			uint8_t dummy[8];
+			spi()->blocking_read(dummy, 8);
+
+			spi()->blocking_read(frame.buffer, length);
+
+			// Receive needs to be dword-aligned, calculate padding bytes needed
+			// and read the padding bytes.
+			uint16_t pad = (-length) & 0x3;
+			if (pad != 0)
+			{
+				spi()->blocking_read(dummy, pad);
+			}
+
+			set_gpio_output_bit(GpioPin::EthCsPin, true);
+
+			register_clear(RegisterOffset::RXQCR, StartDmaAccess);
+			frames_to_process -= 1;
+
+			return std::move(frame);
+		}
+
+		return std::nullopt;
+	}
+
+	/**
+	 * Send a packet.  This will block if no buffer space is available on
+	 * device.
+	 *
+	 * The third argument is a callback that allows the caller to check the
+	 * frame before it's sent but after it's copied into memory that isn't
+	 * shared with other compartments.
+	 */
+	bool send_frame(const uint8_t *buffer, uint16_t length, auto &&check)
+	{
+		// We must check the frame pointer and its length. Although it
+		// is supplied by the firewall which is trusted, the firewall
+		// does not check the pointer which is coming from external
+		// untrusted components.
+		Timeout t{10};
+		if ((heap_claim_fast(&t, buffer) < 0) ||
+		    (!CHERI::check_pointer<CHERI::PermissionSet{
+		       CHERI::Permission::Load}>(buffer, length)))
+		{
+			return false;
+		}
+
+		Debug::log("Sending frame of length {}", length);
+
+		// We would need an intermediate buffer to copy the frame into and
+		// callback to the firewall before sending it.
+		//
+		// Also allocate the space for the start of transmission command (1
+		// byte), header (4 bytes) and the padding (to 4 bytes, not including
+		// the command).
+		uint16_t padded_length  = (length + 3) & ~0x3;
+		auto     transmitBuffer = std::unique_ptr<uint8_t[]>{
+          new (std::nothrow) uint8_t[padded_length + 5]};
+		if (!transmitBuffer)
+		{
+			// Out of memory, drop the frame.
+			return false;
+		}
+
+		// DMA write SPI command.
+		transmitBuffer[0] = 0b11 << 6;
+		uint32_t header   = 0x8000 | (static_cast<uint32_t>(length) << 16);
+		memcpy(&transmitBuffer[1], reinterpret_cast<uint8_t *>(&header), 4);
+		memcpy(&transmitBuffer[5], buffer, length);
+
+		if (!check(&transmitBuffer[5], length))
+		{
+			return false;
+		}
+
+		// Wait for the transmit buffer to be available on the device side.
+		// This needs to include the header.
+		while ((register_read(RegisterOffset::TXMIR) & 0xFFF) < length + 4) {}
+
+		// Start QMU DMA transfer operation.
```

Comment:
The SPI driver is doing PIO instead of DMA to read from SPI, but that's implementation detail of SPI driver and not something that Ethernet driver controls. When you kick start QMU DMA operation, the Ethernet device is DMA directly to the SPI interface and you perform a single SPI transaction, not multiple register reads.

Reply:
Please can you add a comment explaining what DMA means in this context then?